### PR TITLE
Creating TPFs from FFIs

### PR DESCRIPTION
**Pull request by barentsen**
This PR intends to add a feature that lets users cutout a TargetPixelFile from a series of images (e.g. FFIs), inspired by @amcody's [TPFMaker](https://github.com/amcody/TPFmaker/blob/master/TPFmaker.py). 

Example use
--------------
```
    >>> import glob
    >>> import numpy as np
    >>> from astropy.coordinates import SkyCoord
    >>> from lightkurve.targetpixelfile import create_tpf_hdu, KeplerTargetPixelFile

    >>> images = np.sort(glob.glob("/home/gb/proj/k2superstamp/lagoon/*fits"))[0:50]
    >>> position = SkyCoord('18h04m10.10s -24d24m10.4s', frame='icrs')
    >>> hdu = create_tpf_hdu(images, position, size=(10, 10))

    >>> tpf = KeplerTargetPixelFile(hdu)
    >>> lc = tpf.to_lightcurve()
```

Known issues
-----
- [ ] WCS is incorrect
- [ ] TIME values are incorrect (not in BKJD)

### CODE REVIEW

File: lightkurve/targetpixelfile.py [hunk 1]
```diff
@@ -537,3 +543,195 @@ def get_bkg_lightcurve(self, aperture_mask=None):
...
+        # Initialize the 3D data structures
+        self.raw_cnts = np.empty((n_cadences, n_rows, n_cols), dtype='int')
+        self.flux = np.empty((n_cadences, n_rows, n_cols), dtype='float32')
+        self.flux_err = np.empty((n_cadences, n_rows, n_cols), dtype='float32')
+        self.flux_bkg = np.empty((n_cadences, n_rows, n_cols), dtype='float32')
+        self.flux_bkg_err = np.empty((n_cadences, n_rows, n_cols), dtype='float32')
+        self.cosmic_rays = np.empty((n_cadences, n_rows, n_cols), dtype='float32')
+
+        # Set 3D data defaults
+        self.raw_cnts[:, :, :] = -1
```
Comment (mirca) [inline lightkurve/targetpixelfile.py]: You could do: ``self.raw_cnts[:] = -1`` to set all elements of ``self.raw_cnts`` to ``-1`` (independently of the number of dimensions).

File: lightkurve/targetpixelfile.py [hunk 2]
```diff
@@ -537,3 +543,195 @@ def get_bkg_lightcurve(self, aperture_mask=None):
...
+        self.flux[:, :, :] = np.nan
+        self.flux_err[:, :, :] = np.nan
+        self.flux_bkg[:, :, :] = np.nan
+        self.flux_bkg_err[:, :, :] = np.nan
+        self.cosmic_rays[:, :, :] = np.nan
+
+        # Initialize the 1D data structures
+        self.mjd = np.zeros((n_cadences), dtype='float64')
```
Comment (mirca) [inline lightkurve/targetpixelfile.py]: I guess no need for the parenthesis between ``n_cadences``.

File: lightkurve/targetpixelfile.py [hunk 3]
```diff
@@ -537,3 +543,195 @@ def get_bkg_lightcurve(self, aperture_mask=None):
...
+                  'cosmic_rays', 'mjd', 'time', 'timecorr', 'cadenceno',
+                  'quality', 'pos_corr1', 'pos_corr2']
+        for field in fields:
+            if locals()[field] is not None:
+                vars(self)[field][n] = locals()[field]
+
+    def get_tpf_object(self):
+        return KeplerTargetPixelFile(self._make_hdulist())
+
+    def write_fits(self, output_fn=None, overwrite=False):
```
Comment (christinahedges) [inline lightkurve/targetpixelfile.py]: Maybe output_fn should have a sensible default? out.fits, or something yet more sensible?

File: lightkurve/targetpixelfile.py (hunk 1 above)
Comment (christinahedges) [reply lightkurve/targetpixelfile.py]: I actually like that this is easy to read/ see that the variable has 3 dimensions.

File: lightkurve/data/tpf-ext1-header.txt [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+1CUN4P  = 'PIXEL   '           / table column physical WCS axis 1 unit          
+2CUN4P  = 'PIXEL   '           / table column physical WCS axis 2 unit          
+1CRV4P  =                  581 / table column physical WCS ax 1 ref value       
+2CRV4P  =                   90 / table column physical WCS ax 2 ref value       
+1CDL4P  =                  1.0 / table column physical WCS a1 step              
+2CDL4P  =                  1.0 / table column physical WCS a2 step              
+1CRP4P  =                    1 / table column physical WCS a1 reference         
+2CRP4P  =                    1 / table column physical WCS a2 reference         
+WCAX4   =                    2 / number of WCS axes                             
+1CTYP4  = 'RA---TAN'           / right ascension coordinate type
```
Comment (christinahedges) [inline lightkurve/data/tpf-ext1-header.txt]: By using a template header with a WCS in it, are you not artificially adding the wrong WCS to your new data product? Would it be better to have these values as NaNs or 0's in the template just incase they don't get overwritten?

File: lightkurve/targetpixelfile.py [hunk 5]
```diff
@@ -537,3 +543,195 @@ def get_bkg_lightcurve(self, aperture_mask=None):
...
+                    hdu.header[kw] = (self.keywords[kw],
+                                      self.keywords.comments[kw])
+                except KeyError:
+                    hdu.header[kw] = (template[kw],
+                                      template.comments[kw])
+        hdu.header['EXTNAME'] = 'APERTURE'
+        return hdu
+
+
+def create_tpf_hdu(images, position, size=(10, 10), target_id="unnamed-target"):
```
Comment (christinahedges) [inline lightkurve/targetpixelfile.py]: Is `images` a numpy array, fits files or a list of file names? I think it's a list of file names, could we make it a more descriptive name? I would have read that as a numpy array of values or a list of astropy fits objects.

File: lightkurve/data/tpf-ext1-header.txt (hunk 4 above)
Comment (amcody) [reply lightkurve/data/tpf-ext1-header.txt]: I think you're right. When I originally made the template file, I had intended to get rid of the WCS keywords and only write them in later. Maybe we should do that?

**Comment by barentsen** [comment]
I cleaned up the implementation and added both `from_fits_images()` and `to_fits()` methods to `KeplerTargetPixelFile`.  This PR now allows you to cut out TPFs from both K2superstamp and TESS FFI data!!

## K2SUPERSTAMP example
```
import glob
import numpy as np
from astropy.coordinates import SkyCoord
from lightkurve.targetpixelfile import KeplerTargetPixelFile

images = np.sort(glob.glob("/home/gb/proj/k2superstamp/lagoon/*fits"))[0:20]
position = SkyCoord('18h04m10.10s -24d24m10.4s', frame='icrs')
tpf = KeplerTargetPixelFile.from_fits_images(images, position=position, size=(10, 10))
tpf.to_fits("/tmp/my-lagoon-tpf.fits", overwrite=True)
````

## TESS FFI example

```
images = np.sort(glob.glob("/home/gb/proj/tess-simulations/data/ete-6/ffi/2019/132/1-1/*ffic.fits"))[0:20]
position = SkyCoord(254.45420833, 3.35275000, frame='icrs', unit='deg')
tpf = KeplerTargetPixelFile.from_fits_images(images, position=position)
tpf.to_fits("/tmp/my-tess-tpf.fits")
```


## TODO
- [ ] We're not writing the correct WCS to the TPF yet.
- [ ] The TESS example should really produce a `TessTargetPixelFile`.
- [ ] Time is not written in units BKJD.

However none of this is particularly important -- this PR should now be good enough for us to start preparing a notebook to show how TESS FFI images can be mined using lightkurve!

**Comment by mirca** [comment]
> The TESS example should really produce a ``TessTargetPixelFile``

I have yet to code up the ``TessTargetPixelFile`` class, but I wanted to wait for this to get in first =D

**Comment by christinahedges** [comment]
Do we support errors in TPF maker? It would be great to keep them if possible.

**Comment by christinahedges** [comment]
Also could we please support fitsio instead of astropy? It's taking ages to read in all the fits images.

**Comment by barentsen** [comment]
Most of the Keplre cluster images had the  "'Cosmic ray in optimal aperture'" flag set, and were being masked out by the DEFAULT_BITMASK being applied in `KeplerTargetPixelFile`.

I have updated this PR to move the "'Cosmic ray in optimal aperture'" flag from DEFAULT_BITMASK to HARD_BITMASK.  This is a bit risky because I don't fully understand how frequent or important this flag is in non-cluster data, so we may want to revise this decision in the future.

**Comment by barentsen** [comment]
I addded a few more improvements and unit test.  I think the basic functionality is now ready to be merged in time for tess.ninja! Thanks @christinahedges & @gully for testing today.

There are some outstanding improvements for which I am going to open a separate issue:
- The WCS is not carried through yet.
- The performance is slow; using fitsio may be better.
- The unit test coverage is not complete.

**Pull request merged**